File: core/src/io/neilljohnston/github/Player.scala
package io.neilljohnston.github

import com.badlogic.gdx.Gdx
import com.badlogic.gdx.Input.Keys
import com.badlogic.gdx.graphics.Texture
import com.badlogic.gdx.graphics.g2d.{SpriteBatch, TextureRegion}
import io.neilljohnston.github.IWannaJump.Ps
import io.neilljohnston.github.Sprite.KeyControllable

/**
  * Created by Neill on 4/16/2017.
  */
class Player extends FullSprite with KeyControllable with exmath.Smooth {
    // Constants for accel/vel
    final val MagJump: Float = 12 * Ps
    final val MagVel: Float = 8 * Ps
    final val FloatUp: Float = g * 0.45f
    final val FloatDown: Float = g * 0.3f

    // Make player slightly smaller than a tile size
    width = Ps * 0.75f
    height = Ps * 0.75f

    val playerTexture = new Texture(Gdx.files.internal("cat.png"))
    val toDraw = new TextureRegion(playerTexture, 16, 7*16, 16, 16)

    // Add keyboard inputs
    addHoldInput(Keys.LEFT, (d: Float) => if(v.y == 0) v.x += smoothTo(v.x, -MagVel, 20f) * d else v.x += smoothTo(v.x, -MagVel, 8f) * d)
    addHoldInput(Keys.RIGHT, (d: Float) => if(v.y == 0) v.x += smoothTo(v.x, MagVel, 20f) * d else v.x += smoothTo(v.x, MagVel, 8f) * d)
    addTapInput(Keys.SPACE, (_: Float) => if(v.y == 0) v.y = MagJump)
    addHoldInput(Keys.SPACE, (d: Float) => if(v.y > 0) v.y -= FloatUp * d else v.y -= FloatDown * d)
    addRepInput(Keys.DOWN, (d: Float) => v.y = MagJump / 2, 500)

    /**
      * Apply gravity to the sprite.
      * @param delta    Time passed since last step
      */
    override def applyGravity(delta: Float): Unit = {
        if(keyDown(Keys.SPACE))
            v.y = -accelWithMax(-v.y, -gMax * 0.6f, -g * delta)
        else
            v.y = -accelWithMax(-v.y, -gMax, -g * delta)
    }

    /**
      * Override step to include key controls.
      * @param delta    Time passed since last update
      */
    override def step(delta: Float): Unit = {
        callInputs(delta)
        super.step(delta)
        if(!keyDown(Keys.LEFT) && !keyDown(Keys.RIGHT) && v.y == 0)
            v.x *= 0.6f
        if(!keyDown(Keys.LEFT) && !keyDown(Keys.RIGHT) && v.y != 0)
            v.x *= 0.95f
    }

    override def draw(batch: SpriteBatch): Unit = batch.draw(toDraw, x - 8, y - 4, Ps, Ps)
}

File: core/src/io/github/neilljohnston/iwannajump/test/Player.java
package io.github.neilljohnston.iwannajump.test;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.graphics.Texture;
import com.badlogic.gdx.graphics.g2d.Animation;
import com.badlogic.gdx.graphics.g2d.SpriteBatch;
import com.badlogic.gdx.graphics.g2d.TextureRegion;
import com.badlogic.gdx.utils.TimeUtils;
import com.sun.org.apache.xml.internal.resolver.readers.ExtendedXMLCatalogReader;
import io.github.neilljohnston.iwannajump.java.elements.AABB;
import io.github.neilljohnston.iwannajump.java.elements.PlatformCollision;
import io.github.neilljohnston.iwannajump.java.elements.SpriteAABB;
import io.github.neilljohnston.iwannajump.java.elements.SquareCollision;
import io.github.neilljohnston.iwannajump.java.engine.AnimationMap;
import io.github.neilljohnston.iwannajump.java.engine.KeyLogger;
import io.github.neilljohnston.iwannajump.java.math.ExMath;
import io.github.neilljohnston.iwannajump.java.math.Smooth;
import com.badlogic.gdx.Input.Keys;

import static io.github.neilljohnston.iwannajump.java.engine.IWJEnvironment.PS;

public class Player extends SpriteAABB {
    private static final float G = -20 * PS;
    private static final int[] KEYS = {
            Keys.A,
            Keys.D,
            Keys.SPACE,
            Keys.SHIFT_LEFT
    };

    private State state;
    private Direction direction;

    private AnimationMap<State> animations;
    private KeyLogger logger;

    public Player(float x, float y) {
        this.x = x;
        this.y = y;
        this.width = PS * 0.75f;
        this.height = PS * 1.75f;
        this.v.y = -20;

        Texture playerTexture = new Texture(Gdx.files.internal("THE_DUDE.png"));
        animations = new AnimationMap<State>();
        animations.put(State.IDLE, playerTexture, 16, 32, 1f/8f, 0, 0, 0, 0, Animation.PlayMode.LOOP);
        animations.put(State.RUNNING, playerTexture, 16, 32, 1f/8f, 1, 0, 6, 0, Animation.PlayMode.LOOP);
        animations.put(State.WALKING, playerTexture, 16, 32, 1f/8f, 1, 1, 6, 1, Animation.PlayMode.LOOP);

        logger = new KeyLogger(KEYS);

        state = State.IDLE;
        direction = Direction.RIGHT;

        collisions.put(PlatformAABB.class, new PlatformCollision());
    }

    @Override
    public void step(float delta) {
        super.step(delta);

        logger.step(delta);

        v.x *= 0.8;
        state = State.IDLE;
        direction = v.x < 0 ? Direction.LEFT : Direction.RIGHT;

        if(logger.is(Keys.SHIFT_LEFT)) {
            if (logger.is(Keys.A))
                v.x = Smooth.lerp(0, -6.5f * PS, Smooth.unitSin(1f/0.60f * ExMath.absMin(logger.on(Keys.A), 0.60f)));
            if (logger.is(Keys.D))
                v.x = Smooth.lerp(0, 6.5f * PS, Smooth.unitSin(1f/0.60f * ExMath.absMin(logger.on(Keys.D), 0.60f)));

            if(Math.abs(v.x) > PS)
                state = State.RUNNING;
        } else {
            if (logger.is(Keys.A))
                v.x = Smooth.lerp(0, -3.5f * PS, Smooth.unitSin(1f/0.40f * ExMath.absMin(logger.on(Keys.A), 0.40f)));
            if (logger.is(Keys.D))
                v.x = Smooth.lerp(0, 3.5f * PS, Smooth.unitSin(1f/0.40f * ExMath.absMin(logger.on(Keys.D), 0.40f)));

            if(Math.abs(v.x) > PS)
                state = State.WALKING;
        }

        if(v.y == 0) {
            if(logger.is(Keys.SPACE))
                v.y = -0.45f * G;
        } else {
            if(logger.is(Keys.SPACE) && v.y > 0)
                v.y -= 0.30f * G * delta;
        }

        v.y += G * delta;
    }

    @Override
    public void draw(SpriteBatch batch, float delta) {
        Animation<TextureRegion> current = animations.get(state);
        batch.draw(current.getKeyFrame(TimeUtils.millis() % 100000 / 1000f),
                direction == Direction.LEFT ? x + 14 : x - 2, y,
                direction == Direction.LEFT ? -16 : 16, 32);
    }

    private enum State {
        IDLE,
        WALKING,
        RUNNING,
        AIRBORNE
    }

    private enum Direction {
        LEFT,
        RIGHT
    }
}
